### PR TITLE
Optimize comment lookup for group_result by prefetch

### PR DESCRIPTION
**Pull request by okurz**
Same functionality as in before but faster as comments are prefetched for all
jobs in each build in each group. Obviously this could be further optimized
by prefetching comments for all jobs at once but we do not know the jobs
before iterating the builds.

Tested with a production database dump from openqa.suse.de from
2016-02-25 fetching the page 'group_overview/25'. Comparing the number of
queries for comments and the user space time.

Before: 356 @ 2.6s
After: 11 @ 1.9s

*WIP*: It looks like the database from o.s.d is much more demanding and should be used for testing. But the current state of the database seems to have yet another problem so I am not sure against which database dump I should test.

### CODE REVIEW

**Comment by okurz** [comment]
@coolo could help a bit with performance but it is not like I could fully reproduce the problems seen on o.s.d

File: lib/OpenQA/WebAPI/Controller/Main.pm [hunk 1]
```diff
@@ -61,6 +61,12 @@ sub _group_result {
         while (my $line = $keys->next) {
             $settings{$line->job_id}->{$line->key} = $line->value;
         }
+        # prefetch comments to count
+        my $c = $self->db->resultset("Comments")->search({job_id => {in => [map { $_->id } $jobs->all]}});
+        my %labels;
```
Comment (coolo) [inline lib/OpenQA/WebAPI/Controller/Main.pm]: you're not checking labels, you're checking comments here

File: lib/OpenQA/WebAPI/Controller/Main.pm [hunk 2]
```diff
@@ -81,7 +87,7 @@ sub _group_result {
                     || $job->result eq OpenQA::Schema::Result::Jobs::INCOMPLETE)
                 {
                     $jr{failed}++;
-                    $jr{labeled}++ if $job->comments > 0;
+                    $jr{labeled}++ if ($labels{$job->id} && $labels{$job->id} > 0);
```
Comment (coolo) [inline lib/OpenQA/WebAPI/Controller/Main.pm]: you still don't trust perl's if? It can't get < 0

File: lib/OpenQA/WebAPI/Controller/Main.pm (hunk 1 above)
Comment (aaannz) [inline lib/OpenQA/WebAPI/Controller/Main.pm]: Isn't label comment just in special format?
In [Test.pm:345](https://github.com/os-autoinst/openQA/blob/master/lib/OpenQA/WebAPI/Controller/Test.pm#L345) `%labels` include comments too.

**Comment by okurz** [comment]
rebased and measured again with database dump from o.s.d:
```
$ echo "after"; avgtime -r5 -q sh -c "for url in '/tests/overview?distri=sle&version=12-SP2&build=1171&groupid=25' '/' '/group_overview/25' ; do export date=2016-02-25; sudo -u geekotest OPENQA_SQL_DEBUG=1 OPENQA_CONFIG=local/stage_openqa_suse_de_${date} OPENQA_DATABASE=stage script/openqa get $url 'foobar' ; done"
after

------------------------
Total time (ms): 30678.1
Repetitions    : 5
Sample mode    : 6220 (1 occurrences)
Median time    : 6139.42
Avg time       : 6135.61
Std dev.       : 58.915
Minimum        : 6055.84
Maximum        : 6224.72
95% conf.int.  : [6020.14, 6251.08]  e = 115.471
99% conf.int.  : [5983.86, 6287.37]  e = 151.755
EstimatedAvg95%: [6083.97, 6187.25]  e = 51.6404
EstimatedAvg99%: [6067.75, 6203.48]  e = 67.8669
okurz@linux-28d6:~/local/os-autoinst/openQA 0 (enhance/comments_queries) $ echo "before"; avgtime -r5 -q sh -c "for url in '/tests/overview?distri=sle&version=12-SP2&build=1171&groupid=25' '/' '/group_overview/25' ; do export date=2016-02-25; sudo -u geekotest OPENQA_SQL_DEBUG=1 OPENQA_CONFIG=local/stage_openqa_suse_de_${date} OPENQA_DATABASE=stage script/openqa get $url 'foobar' ; done"
before

------------------------
Total time (ms): 30767.6
Repetitions    : 5
Sample mode    : 6290 (1 occurrences)
Median time    : 6127.31
Avg time       : 6153.52
Std dev.       : 76.3568
Minimum        : 6074.94
Maximum        : 6295.14
95% conf.int.  : [6003.87, 6303.18]  e = 149.657
99% conf.int.  : [5956.84, 6350.2]  e = 196.682
EstimatedAvg95%: [6086.59, 6220.45]  e = 66.9285
EstimatedAvg99%: [6065.56, 6241.48]  e = 87.9589
```
-> 6135+-59 vs. 6153+-76 ==> **no improvement** performance wise

Still, can go ahead and merge it if you prefer this as the "cleaner" way, I am not convinced.

File: lib/OpenQA/WebAPI/Controller/Main.pm (hunk 2 above)
Comment (okurz) [inline lib/OpenQA/WebAPI/Controller/Main.pm]: changed to single if

File: lib/OpenQA/WebAPI/Controller/Main.pm (hunk 1 above)
Comment (okurz) [inline lib/OpenQA/WebAPI/Controller/Main.pm]: Added comment to describe why "labels =~ comments"

**Comment by coolo** [comment]
but the log will be much cleaner

**Comment by okurz** [comment]
but the code might not :-/

**Comment by okurz** [comment]
merge if you like

**Pull request merged**